### PR TITLE
Add richer structure for Stable MIR Projections

### LINKED ISSUE #49: Add support for Projection
Projection is currently encoded as `String`, which doesn't provide any insight for the user and makes `Place` fairly unusable.

### PR DESCRIPTION
**Pull request by klinvill**
Resolves https://github.com/rust-lang/project-stable-mir/issues/49. 

Projections in Stable MIR are currently just strings. This PR replaces that representation with a richer structure, namely projections become vectors of `ProjectionElem`s, just as in MIR. The `ProjectionElem` enum is heavily based off of the MIR `ProjectionElem`.

This PR is a draft since there are several outstanding issues to resolve, including:

- How should `UserTypeProjection`s be represented in Stable MIR? In MIR, the projections are just a vector of `ProjectionElem<(),()>`, meaning `ProjectionElem`s that don't have Local or Type arguments (for `Index`, `Field`, etc. objects). Should `UserTypeProjection`s be represented this way in Stable MIR as well? Or is there a more user-friendly representation that wouldn't drag along all the `ProjectionElem` variants that presumably can't appear?
- What is the expected behavior of a `Place`'s `ty` function? Should it resolve down the chain of projections so that something like `*_1.f` would return the type referenced by field `f`?
- Tests should be added for `UserTypeProjection`

### CODE REVIEW

**Comment by celinval** [comment]
@klinvill I'm not an official reviewer, I'll take a look at it though.

r? @spastorino

can you take a look at it please?

**Review by celinval — COMMENTED** [review]
I would like to suggest that you crate a new test instead of extending the existing one. Thanks for doing this though!

File: compiler/stable_mir/src/mir/body.rs [hunk 1]
```diff
@@ -533,6 +644,8 @@ impl Constant {
 }
 
 impl Place {
+    // TODO(klinvill): What is the expected behavior of this function? Should it resolve down the
```
[the remark this answers is not in the document]
Comment (klinvill) [reply compiler/stable_mir/src/mir/body.rs]: Sounds good, I'll defer this to a separate PR

**Comment by ouz-a** [comment]
> 
>     * Within `ProjectionElem`, the `Downcast` variant in MIR includes an `Option<Symbol>` argument that can contain the name of the variant and comments state that it is used for printing the MIR. However, I believe this information can be obtained just using the `VariantIdx` argument. I've therefore omitted the `Option<Symbol>` argument. Is there instead a reason to keep it around?
> 
>     * I haven't been able to generate MIR that contains the `UserTypeProjection` struct. Is it still possible to generate `UserTypeProjection` MIR after type ascriptions were removed in accordance with [De-RFC: Remove type ascription rfcs#3307](https://github.com/rust-lang/rfcs/pull/3307)? If so, in what cases is this struct generated? In the meantime I've left the projections field in `UserTypeProjection` unimplemented (specifically it's still a string).
> 
>     * In MIR, the projections in `UserTypeProjection` and `Place` use different instantiations of the same generic enum `ProjectionElem`. In particular, it seems like `UserTypeProjection` projections only support a subset of the variants. I've therefore opted to hardcode `ProjectionElem` to the instantiation for `Place`, and will create a separate enum if `UserTypeProjection`s should still be included in Stable MIR.
> 
> 

I will try to address these.

1. Imo we should always try to have printable version of a thing available for error reporting and debug printing also this will be useful for https://github.com/rust-lang/rust/pull/117745 once it lands
2/3. I think we can leave them for now and if we hit `UserTypeProjection` we know we have a case for it and we can implement it. This can be done randomly testing stuff in `ui`

**Comment by celinval** [comment]
Hi @klinvill, I think it makes sense to remove the `Option<Symbol>` as long as we provide an utility function that retrieves that information. Otherwise, we might want to keep it as pointed out by @ouz-a.

Can you change the `projection` field of `UserTypeProjection` to `Opaque` type for now? Thanks

**Comment by klinvill** [comment]
Thanks for the suggestions @ouz-a and @celinval. The inconsistency of having `Option<Symbol>` only to go along with the `VariantIdx` in `Downcast`, and not having a similar field to go with `FieldIdx` in `Field` bugs me a bit. I think it would be better long-term to be consistent with either passing around the additional information, or having that information accessible through a lookup based on the index.

I'd advocate for having that information accessible through a lookup based on the index rather than passing around additional information (which would potentially need to be updated as Variants and Fields evolve). However, such a lookup is currently neither possible (due to `AdtDef` not being implemented) nor convenient (due to `ty` not being implemented for `Place`). Currently, looking up the variant in a projection element would require accessing the local field in the parent `Place` object, then using that (index) to locate the appropriate `LocalDecl`, then following down the projection elements that precede the `Downcast` to get the appropriate `AdtDef`, then retrieving the proper `VariantDef` from the variants in the `AdtDef` (which currently isn't implemented). Ideally, we could instead create a copy of the `Place` definition with the projection elements truncated to either the desired `Downcast` projection or its parent, and then just call the `ty` function to get the appropriate `AdtDef`, but the `ty` function also isn't implemented yet.

So I suppose this is my long-winded way of proposing to drop the `Option<Symbol>` field with the intention of having utility functions that can do the lookup based on the index. However, since it's currently not possible to do this, I propose deferring the utility functions (at least for variant lookups for `Downcast`) until `AdtDef` and `Place::ty` are implemented. Does that sound reasonable to you both @ouz-a and @celinval?

**Comment by celinval** [comment]
Sounds reasonable to me. Can you please capture your comment above into an issue in [project-stable-mir](https://github.com/rust-lang/project-stable-mir)? We should revisit this in the near term, and create a follow up PR.

Thank you

**Comment by ouz-a** [comment]
This is probably nothing 
@bors r+ rollup

**Comment by ouz-a** [comment]
Maybe we should re-try this again before rolling it up
@bors r-

**Pull request merged**